src/ImageSharp/Diagnostics/MemoryDiagnostics.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Threading;
+
+namespace SixLabors.ImageSharp.Diagnostics
+{
+    /// <summary>
+    /// Represents the method to handle <see cref="MemoryDiagnostics.UndisposedAllocation"/>.
+    /// </summary>
+    public delegate void UndisposedMemoryResourceDelegate(string allocationStackTrace);
+
+    /// <summary>
+    /// Utilities to track memory usage and detect memory leaks from not disposing ImageSharp objects.
+    /// </summary>
+    public static class MemoryDiagnostics
+    {
+        private static int totalUndisposedAllocationCount;
+
+        private static UndisposedMemoryResourceDelegate undisposedMemoryResource;
+        private static int undisposedMemoryResourceSubscriptionCounter;
+
+        /// <summary>
+        /// Fires when an ImageSharp object's undisposed memory resource leaks to the finalizer.
+        /// The event brings significant overhead, and is intended to be used for troubleshooting only.
+        /// For production diagnostics, use <see cref="TotalUndisposedAllocationCount"/>.
+        /// </summary>
+        public static event UndisposedMemoryResourceDelegate UndisposedAllocation
+        {
+            add
+            {
+                Interlocked.Increment(ref undisposedMemoryResourceSubscriptionCounter);
+                undisposedMemoryResource += value;
+            }
+
+            remove
+            {
+                undisposedMemoryResource -= value;
+                Interlocked.Decrement(ref undisposedMemoryResourceSubscriptionCounter);
+            }
+        }
+
+        /// <summary>
+        /// Gets a value indicating the total number of memory resource objects leaked to the finalizer.
+        /// </summary>
+        public static int TotalUndisposedAllocationCount => totalUndisposedAllocationCount;
+
+        internal static bool MemoryResourceLeakedSubscribed => undisposedMemoryResourceSubscriptionCounter > 0;
+
+        internal static void IncrementTotalUndisposedAllocationCount() =>
+            Interlocked.Increment(ref totalUndisposedAllocationCount);
+
+        internal static void DecrementTotalUndisposedAllocationCount() =>
+            Interlocked.Decrement(ref totalUndisposedAllocationCount);
+
+        internal static void RaiseUndisposedMemoryResource(string allocationStackTrace)
+        {
+            // Schedule on the ThreadPool, to avoid user callback messing up the finalizer thread.
+            ThreadPool.QueueUserWorkItem(_ => undisposedMemoryResource?.Invoke(allocationStackTrace));
```

Comment:
Perform the null-check before queuing to thread-pool? So if nothing is registered, the thread-pool has less work to do. Ala
```suggestion
            UndisposedMemoryResourceDelegate tmp = undisposedMemoryResource;
            if (tmp is not null)
            {
                ThreadPool.QueueUserWorkItem(static state =>
                {
                    var (@event, stackTrace) = state;
                    @event(stackTrace);
                }, (tmp, allocationStackTrace), preferLocal: false);
            }
```
This also avoids the closures for the stacktrace and the delegate, thus avoiding allocations*.
Though as noted in the PR's description this API is for diagnostics and not intended for production, so perf doesn't matter as much one could go a simpler route:
```suggestion
            UndisposedMemoryResourceDelegate tmp = undisposedMemoryResource;
            if (tmp is not null)
            {
                ThreadPool.QueueUserWorkItem(stackTrace => tmp(stackTrace), allocationStackTrace, preferLocal: false);
            }
```

Anyway my main point is: cache to local, perform the null-check, and iif not null queue to thread-pool to have minimal impact for production.

---

\* in thread-pool internals there's still an allocation happening, which could be avoided by a type that implements `IThreadPoolWorkItem` which can be re-used. But that's overkill here.

Reply:
>  in thread-pool internals there's still an allocation happening, which could be avoided by a type that implements IThreadPoolWorkItem which can be re-used. But that's overkill here.

Yeah I considered that, but definitely not worth it. Users should not subscribe to this delegate in production anyways, it takes a snapshot of `Environment.StackTrace` on *every* buffer allocation.